src/ProxyProtocol.cc

```diff
@@ -0,0 +1,306 @@
+        throw TexcHere("PROXY/2.0 error: invalid command");
+
+    char rawFamily = tok.uint8("family");
+
+    const char family = (rawFamily & 0xF0) >> 4;
+    if (family > 0x3) // values other than 0x0-0x3 are invalid
+        throw TexcHere("PROXY/2.0 error: invalid family");
+
+    const char proto = (rawFamily & 0x0F);
+    if (proto > 0x2) // values other than 0x0-0x2 are invalid
+        throw TexcHere("PROXY/2.0 error: invalid protocol type");
+
+    const uint16_t headerLen = tok.uint16("length") + mandatoryHeaderLen;
+
+    if (buf_.length() < headerLen)
+        return false; // need more bytes
+
+    // LOCAL connections do nothing with the extras
+    if (command == 0x00/* LOCAL*/)
+        return true;
+
+    union pax {
+        struct {        /* for TCP/UDP over IPv4, len = 12 */
+            struct in_addr src_addr;
+            struct in_addr dst_addr;
+            uint16_t src_port;
+            uint16_t dst_port;
+        } ipv4_addr;
+        struct {        /* for TCP/UDP over IPv6, len = 36 */
+            struct in6_addr src_addr;
+            struct in6_addr dst_addr;
+            uint16_t src_port;
+            uint16_t dst_port;
+        } ipv6_addr;
+        struct {        /* for AF_UNIX sockets, len = 216 */
+            uint8_t src_addr[108];
+            uint8_t dst_addr[108];
+        } unix_addr;
+    };
+
+    switch (family) {
+
+    case 0x1:  { // IPv4
+        pax ipu;
+        const SBuf rawAddr = tok.area(sizeof(pax::ipv4_addr), "ipv4_addr");
+        memcpy(&ipu, rawAddr.rawContent(), rawAddr.length());
+
+        dstIpAddr = ipu.ipv4_addr.dst_addr;
+        dstIpAddr.port(ntohs(ipu.ipv4_addr.dst_port));
+        srcIpAddr = ipu.ipv4_addr.src_addr;
+        srcIpAddr.port(ntohs(ipu.ipv4_addr.src_port));
+        break;
+    }
+
+    case 0x2:  { // IPv6
+        pax ipu;
+        const SBuf rawAddr = tok.area(sizeof(pax::ipv6_addr), "ipv6_addr");
+        memcpy(&ipu, rawAddr.rawContent(), rawAddr.length());
+
+        dstIpAddr = ipu.ipv6_addr.dst_addr;
+        dstIpAddr.port(ntohs(ipu.ipv6_addr.dst_port));
+        srcIpAddr = ipu.ipv6_addr.src_addr;
+        srcIpAddr.port(ntohs(ipu.ipv6_addr.src_port));
+        break;
+    }
+
+    case 0x3:  { // TODO: add support for AF_UNIX sockets.
+        tok.skip(sizeof(pax::unix_addr), "unix_addr");
+        break;
+    }
+
+    default: {
+        // the invalid family, we have checked already
+        assert(0);
+        break;
+    }
+    }
+
+    while (tok.parsed() < headerLen) { 
+        assert(!tok.atEnd());
+        if (!v2Message)
+            v2Message = new ProxyProtocol::Two::Message;
+        ProxyProtocol::Two::Tlv tlv(tok.uint8("pp2_tlv::type"));
+        const uint16_t valueLen = tok.uint16("pp2_tlv::length");
+        if (tok.parsed() + valueLen > headerLen)
+            throw TexcHere("PROXY/2.0 error: invalid pp2_tlv length");
+        tlv.value = tok.area(valueLen, "pp2_tlv::value");
+        v2Message->tlvs.emplace_back(tlv);
```

Comment:
The `emplace_back` method is meant for avoiding an (extra) object creation. You have already created the object so this call is out of place here. In other words, emplace_back() is meant to be used with regular constructor arguments, not a copy constructor argument (it is that copying that the method is meant to avoid!).

This problem will be resolved if you supply the type and value arguments here, without creating a Tlv object (as discussed earlier) instead of abusing the copy constructor in an emplace_back() call.